### PR TITLE
Implement blinded units attack penalties

### PR DESCRIPTION
**Pull request by vmarkovtsev**
Last TODO

### CODE REVIEW

File: lib/CBattleCallback.cpp [hunk 1]
```diff
@@ -1020,8 +1020,28 @@ TDmgRange CBattleInfoCallback::calculateDmgRange(const BattleAttackInfo &info) c
 		multBonus *= 0.5;
 	}
 
-	// TODO attack on petrified unit 50%
 	// blinded unit retaliates
+	auto blinded = info.attackerBonuses->getBonus(Selector::source(Bonus::SPELL_EFFECT, SpellID::BLIND));
```
Comment (alexvins) [inline lib/CBattleCallback.cpp]: blind spell has 3 effects = 3 bonuses (2 of them with 0 val). Which bonus will you get?

File: lib/CBattleCallback.cpp [hunk 2]
```diff
@@ -1020,8 +1020,28 @@ TDmgRange CBattleInfoCallback::calculateDmgRange(const BattleAttackInfo &info) c
...
 	}
 
-	// TODO attack on petrified unit 50%
 	// blinded unit retaliates
+	auto blinded = info.attackerBonuses->getBonus(Selector::source(Bonus::SPELL_EFFECT, SpellID::BLIND));
+	if(blinded)
+	{
+		switch(blinded->val)
+		{
+			case SecSkillLevel::BASIC:
```
Comment (alexvins) [inline lib/CBattleCallback.cpp]: Spell effect bonus "val" can`be SecSkillLevel. It is explicitly defined number (defaults to "power" spell attribute).

Comment (vmarkovtsev) [inline lib/CBattleCallback.cpp]: I looked at
```
auto isAdvancedAirShield = [](const Bonus *bonus)
	{
		return bonus->source == Bonus::SPELL_EFFECT
			&& bonus->sid == SpellID::AIR_SHIELD
			&& bonus->val >= SecSkillLevel::ADVANCED;
	};
```
below. Is that code wrong too?

File: lib/CBattleCallback.cpp (hunk 1 above)
Comment (vmarkovtsev) [inline lib/CBattleCallback.cpp]: So the TODO is outdated and everything works as-is, right?

File: lib/CBattleCallback.cpp (hunk 2 above)
Comment (DjWarmonger) [inline lib/CBattleCallback.cpp]: Yes it's wrong altogether.

File: lib/CBattleCallback.cpp (hunk 1 above)
Comment (alexvins) [inline lib/CBattleCallback.cpp]: Need testing. Have not tested this exact case.

File: lib/CBattleCallback.cpp (hunk 2 above)
Comment (DjWarmonger) [inline lib/CBattleCallback.cpp]: Air shield gives GENERAL_DAMAGE_REDUCTION bonus.
This piece of code effectively does nothing and is outdated.

On the other hand, the code indicates that distance penalty and air shield penalty should not be cumulative. Strategija reads:
"Attacker is a shooter with range, wall or melee penalty, or Advanced Air Shield is cast"

Though I'm not sure it's correct or logical.

**Comment by vmarkovtsev** [comment]
Closing as this is already implemented

**Pull request closed without merging**